### PR TITLE
Pilot: add protocol sniffing flag for inbound listener

### PR DESCRIPTION
**Pull request by yxue**
Add protocol sniffing flag for inbound listener. Before there was a PR for adding flag for virtual inbound listener. 

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 1]
```diff
@@ -364,7 +364,8 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(
 			}
 
 			pluginParams := &plugin.InputParams{
-				ListenerProtocol:           plugin.ModelProtocolToListenerProtocol(node, endpoint.ServicePort.Protocol),
+				ListenerProtocol: plugin.ModelProtocolToListenerProtocol(node, endpoint.ServicePort.Protocol,
```
Comment (lambdai) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: I remember during our offline chat you mentioned the inbound filter chain match case 2, which is TLS + TCP, you changed from alpn=[istio] to alpn=[].
Is here the key to revert the change?

**Review by lambdai — COMMENTED** [review]
It's a little scary that no test case is added. At least add unit test for ModelProtocolToListenerProtocol?

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 1 above)
Comment (yxue) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: This PR is for add flag for adding missing flag for inbound listener only.

File: pilot/pkg/networking/plugin/plugin.go [hunk 2]
```diff
@@ -51,10 +52,22 @@ const (
...
-	if !util.IsProtocolSniffingEnabledForNode(node) && p == protocol.Unsupported {
-		p = protocol.TCP
+	if p == protocol.Unsupported {
+		switch trafficDirection {
+		case core.TrafficDirection_INBOUND:
+			if !util.IsProtocolSniffingEnabledForInbound(node) {
+				p = protocol.TCP
+			}
+		case core.TrafficDirection_OUTBOUND:
+			if !util.IsProtocolSniffingEnabledForNode(node) {
```
Comment (rshriram) [inline pilot/pkg/networking/plugin/plugin.go]: dont you want to do a search and replace and call this IsProtocolSniffingEnabledForOutbound ?

Comment (yxue) [reply pilot/pkg/networking/plugin/plugin.go]: Renamed

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 1 above)
Comment (lambdai) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: If the flag can fully turn off the inbound listener I am fine.

File: install/kubernetes/helm/istio/charts/pilot/values.yaml [hunk 3]
```diff
@@ -12,8 +12,10 @@ rollingMaxUnavailable: 25%
 image: pilot
 sidecar: true
 traceSampling: 1.0
-# if protocol sniffing is enabled. Default to true.
-enableProtocolSniffing: true
+# if protocol sniffing is enabled for outbound. Default to true.
```
Comment (howardjohn) [inline install/kubernetes/helm/istio/charts/pilot/values.yaml]: nit - Just remove the `default to...`. If they are reading this they can see the default (and its not accurate :slightly_smiling_face: )

**Review by howardjohn — APPROVED** [review]

File: install/kubernetes/helm/istio/charts/pilot/values.yaml [hunk 4]
```diff
@@ -12,8 +12,10 @@ rollingMaxUnavailable: 25%
 image: pilot
 sidecar: true
 traceSampling: 1.0
-# if protocol sniffing is enabled. Default to true.
-enableProtocolSniffing: true
+# if protocol sniffing is enabled for outbound
+enableProtocolSniffingForOutbound: true
+# if protocol sniffing is enabled for inbound
+enableProtocolSniffingForInbound: false
```
Comment (rshriram) [inline install/kubernetes/helm/istio/charts/pilot/values.yaml]: why is this being turned off?

Comment (yxue) [reply install/kubernetes/helm/istio/charts/pilot/values.yaml]: There is a bug when destination rule is applied. The inbound cannot inspect ALPN when TLS is enabled.

**Review by rshriram — CHANGES_REQUESTED** [review]
please hold. Can you point to the issue that is causing you to turn off sniffing for inbound? 
I also realized that while we can do inbound/outbound in code, the telemetry will be badly broken as it will be only one way. And if someone is forced to name their ports for inbound telemetry,  its as good as turning off sniffing because people are now being forced to name all ports!

**Comment by lambdai** [comment]
short version:
Context:
Suppose user enable mtls, at outbound listener, envoy could sniff out whether it's http or tcp. However, the outbound filter chain will always apply the same tls_context where alpn is ["istio"]
Risk:
At inbound, there is no approach to determine the the protocol undermined by tls: http_tls doesn't work with tls traffic, alpn from tls_inspector are always "istio" as per above context.
Currently traffic would match http and the bytes are consumed by http connection manager, even though the original traffic is TCP(expecting tcp_proxy instead of http conn manager)

**Comment by linsun** [comment]
@lambdai - thank you. If I can replay the risk, when mTLS stict is enabled, the secure communication among 2 services (Service A -> B) could be broken due to unable to detect the correct protocol (inbound for service B).  Is this correct?

If we enable this PILOT_ENABLE_PROTOCOL_SNIFFING_FOR_INBOUND = false , does it mean user will need to manually name inbound port as they have today in service YAML file?

**Comment by yxue** [comment]
@linsun 

If permissive mTLS is enabled. If incoming traffic is plain text traffic (TCP or HTTP), everything is good. If incoming traffic TLS traffic, for HTTP traffic, it's good; for TCP traffic, it will hit http filter chain and be broken.

If strict mTLS is enabled. For TCP traffic, it's good; for http traffic, it will hit tcp filter chain and be downgraded.

> If we enable this PILOT_ENABLE_PROTOCOL_SNIFFING_FOR_INBOUND = false , does it mean user will need to manually name inbound port as they have today in service YAML file?

Yes

**Comment by rshriram** [comment]
So here is the problem as I understand it. When the incoming traffic is mTLS, the http inspector will not be able to determine whether it is http or tcp traffic. It doesn't matter if we have mtls strict or permissive.

And in this scenario, due to certain deficiencies in way filter chain matching works in Envoy, it ends up matching a HTTP filter chain and messes up TCP traffic. It passed our e2e tests because all our e2e workloads only send/receive HTTP traffic [even though the ports are marked as TCP] due to convenience and ease of testing. 

A proper solution would involve the following changes in Envoy IMO:
 1. on the sender side, we need to store the alpn to send based on the sniffing outcome in the outbound listener. This alpn needs to be added to the connection when making upstream mTLS connections via cluster.
 2. on receiver side, add a `alpn_match_all` boolean to Envoy, and possibly multiple alpn sets in the filter chain. We would then generate a filter chain match with
 ```
   - alpnSets:
      - istio, http/1.1
      - istio, h2
      - istio, http/1.0
     transportProtocol: TLS
    //// add Http connection manager here
  - alpnSets:
     - istio
    transportProtocol: TLS
   /// add TCP proxy here
```

and then followed by appropriate pilot level changes.

**Comment by rlenglet** [comment]
This sounds like significantly impactful work. I believe this should require a proper design doc and a proper test plan.
I prefer _not_ to hold the 1.3.0 release until this is done.
Can we please merge this PR to disable this feature by default and ship 1.3.0, _then_ figure out a plan?

**Comment by rshriram** [comment]
Yes, but I first want to get some ack from @yxue / @lambdai that this is the issue and that there are no other hidden gotchas here.

**Review by rshriram — APPROVED** [review]

**Comment by yxue** [comment]
@rshriram 

The first solution is what we are going to do. There are some problems with the second one because two filter chains will conflict with each other. But I think similar solution can be applied here. e.g., use `istio-http/1.1` instead of `[istio, http/1.1]`.

**Comment by lambdai** [comment]
step 1: an easy impl from envoy side: https://github.com/envoyproxy/envoy/issues/8197
step 2: It's kind of breaking change. Need more discussion on details

**Comment by incfly** [comment]
Jump in late, whatever the changes we're making, we need to ensure the PERMISSIVE mode will be updated correspondingly, using new ALPN values to match the mTLS chain. istio/xx rather than "istio".
